src/ucp/core/ucp_request.c

```diff
@@ -444,29 +413,26 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_request_memory_reg,
         }
 
         for (iov_it = 0; iov_it < iovcnt; ++iov_it) {
-            if (update) {
-                if (ucs_test_all_flags(memhs[iov_it]->md_map, reg_md_map)) {
-                    continue;
-                }
-                status = ucp_memh_update(context, reg_md_map, flags,
-                                         memhs[iov_it]);
-            } else {
-                status = ucp_memh_get(context, iov[iov_it].buffer,
-                                      iov[iov_it].length, mem_type, reg_md_map,
-                                      flags, &memhs[iov_it]);
-            }
+            status = ucp_memh_update(context, iov[iov_it].buffer,
+                                     iov[iov_it].length, mem_type, reg_md_map,
+                                     flags, &memhs[iov_it]);
             if (status != UCS_OK) {
                 /* unregister previously registered memory */
-                ucp_request_dt_dereg(context, memhs, iov_it, req);
+                if (state->dt.iov.memhs == NULL) {
+                    ucp_request_dt_dereg(context, memhs, iov_it, req);
+                } else {
+                    ucp_request_dt_dereg(context, memhs, iovcnt, req);
+                    state->dt.iov.memhs = NULL;
+                }
```

Comment:
is this condition for proto_v2? then better to check for `context->config.ext.proto_enable` or `req->flags & UCP_REQUEST_FLAG_PROTO_SEND`